python/ml4ir/base/data/relevance_dataset.py

```diff
@@ -25,16 +27,26 @@ def __init__(
         test_pcent_split: float = -1,
         use_part_files: bool = False,
         parse_tfrecord: bool = True,
-        logger: Logger = None,
+        logger: Optional[Logger] = None,
     ):
         self.feature_config = feature_config
         self.max_sequence_size = max_sequence_size
-        self.data_dir: str = data_dir
+        self.logger = logger
+
+        # If data directory is a HDFS path, first copy to local file system
```

Comment:
I think maybe we can reorganize this a bit.  This has references to sparkIO directly in this class, which we don't want.  We should encapsulate the IO into a class (ahh! Jake is suggesting a class?!?) or a function which is passed in.

We can chat about that later on if it's not clear.

Reply:
(Haha. I went overboard with NOT making spark_io a class this time.)
I'm not sure how we can eliminate a reference to spark_io here though. I can wrap it under `file_io.copy_dir(src, dest)` and then internally call `spark_io` like I do for the other I/O codepaths, if you like.

Does that work?

Reply:
what I mean is you could pass into RelevanceDataset's constructor, an object or function which provides IO.  If it's a object which you pass in, the (abstract) class of the object would be DataIO (or something like that), and some particular implementations would be LocalIO and SparkIO.  Then RelevanceDataset has no references to spark, but whoever instantiates a RelevanceDataset may choose to create a spark-based DataIO and pass it into the constructor.

Reply:
I understand what you are suggesting. But I don't get how you would do this specific task - "copying files from HDFS to Local filesystem", which is specific for the spark_io case.

Reply:
Keep in mind that for tfrecord, we don't explicitly(using file_io or spark_io) do any read at all. It's handled by tensorflow.

Reply:
Alternatively, I can do the copying in the main pipeline and force RelevanceDataset to work with only file_io(or local file I/O). This might be cleaner. My reasoning for adding the copying to RelevanceDataset was that if a user wanted to create a RelevanceDataset using data stored on HDFS, then they already can do it. They don't have to write their own version of spark_IO. May be passing in the file handler object into RelevanceDataset is a good middle ground. Not entirely convinced.